nitro-node/src/index.ts

```diff
@@ -0,0 +1,348 @@
+ */
+async function killSubprocess(): Promise<void> {
+  const controller = new AbortController();
+  setTimeout(() => controller.abort(), 5000);
+  log(`[NITRO]::Debug: Request to kill Nitro`);
+
+  return fetch(NITRO_HTTP_KILL_URL, {
+    method: "DELETE",
+    signal: controller.signal,
+  })
+    .then(() => {
+      subprocess?.kill();
+      subprocess = undefined;
+    })
+    .catch(() => { })
+    .then(() => tcpPortUsed.waitUntilFree(PORT, 300, 5000))
+    .then(() => log(`[NITRO]::Debug: Nitro process is terminated`));
+}
+
+/**
+ * Spawns a Nitro subprocess.
+ * @returns A promise that resolves when the Nitro subprocess is started.
+ */
+function spawnNitroProcess(): Promise<void> {
+  log(`[NITRO]::Debug: Spawning Nitro subprocess...`);
+
+  return new Promise<void>(async (resolve, reject) => {
+    const binaryFolder = path.join(__dirname, "..", "bin"); // Current directory by default
+    const executableOptions = executableNitroFile();
+
+    const args: string[] = ["1", LOCAL_HOST, PORT.toString()];
+    // Execute the binary
+    log(
+      `[NITRO]::Debug: Spawn nitro at path: ${executableOptions.executablePath}, and args: ${args}`
+    );
+    subprocess = spawn(
+      executableOptions.executablePath,
+      ["1", LOCAL_HOST, PORT.toString()],
+      {
+        cwd: binaryFolder,
+        env: {
+          ...process.env,
+          CUDA_VISIBLE_DEVICES: executableOptions.cudaVisibleDevices,
+        },
+      }
+    );
+
+    // Handle subprocess output
+    subprocess.stdout.on("data", (data: any) => {
+      log(`[NITRO]::Debug: ${data}`);
+    });
+
+    subprocess.stderr.on("data", (data: any) => {
+      log(`[NITRO]::Error: ${data}`);
+    });
+
+    subprocess.on("close", (code: any) => {
+      log(`[NITRO]::Debug: Nitro exited with code: ${code}`);
+      subprocess = undefined;
+      reject(`child process exited with code ${code}`);
+    });
+
+    tcpPortUsed.waitUntilUsed(PORT, 300, 30000).then(() => {
+      log(`[NITRO]::Debug: Nitro is ready`);
+      resolve();
+    });
+  });
+}
+
+/**
+ * Get the system resources information
+ * TODO: Move to Core so that it can be reused
+ */
+function getResourcesInfo(): Promise<ResourcesInfo> {
+  return new Promise(async (resolve) => {
+    const cpu = osUtils.cpuCount();
+    log(`[NITRO]::CPU informations - ${cpu}`);
+    const response: ResourcesInfo = {
+      numCpuPhysicalCore: cpu,
+      memAvailable: 0,
+    };
+    resolve(response);
+  });
+}
+
+/**
+ * Every module should have a dispose function
+ * This will be called when the extension is unloaded and should clean up any resources
+ * Also called when app is closed
+ */
+function dispose() {
+  // clean other registered resources here
+  killSubprocess();
+}
+
+export default {
+  setLogger,
+  runModel,
+  stopModel,
+  killSubprocess,
+  dispose,
+  updateNvidiaInfo,
```

Comment:
The module should not depend on the app's business logic. Let's assume that the app's setting folder directory might be different between apps, which could confuse users. The module should be stateless or sandboxed & the folder might not exist.